### PR TITLE
Add two versions of Simrank similarity

### PR DESCRIPTION
**Pull request by mrecachinas**
This PR supersedes #1691 as it pulls in changes (and fixes merge conflicts) from the baseline since 2015 and it also includes tests.

Furthermore, I’ve added a pure numpy matrix calculation of SimRank as another function available for use in similarity.py.

cc @jfinkels @hagberg

### CODE REVIEW

File: networkx/algorithms/similarity.py [hunk 1]
```diff
@@ -1007,6 +1010,211 @@ def prune(cost):
...
+    if not isinstance(d1, dict) and not isinstance(d2, dict):
+        return abs(d1 - d2) <= atolerance + rtolerance * abs(d2)
+    return all(all(_is_close(d1[u][v], d2[u][v]) for v in d1[u]) for u in d1)
+
+
+def simrank_similarity(G, source=None, target=None, importance_factor=0.9,
+                       max_iterations=100, tolerance=1e-4):
+    """Returns the SimRank similarity of nodes in the graph ``G``.
+
+    The SimRank algorithm for determining node similarity is defined in
```
[the remark this answers is not in the document]
Comment (mrecachinas) [reply networkx/algorithms/similarity.py]: Addressed in 6d6ff62.

File: networkx/algorithms/similarity.py [hunk 2]
```diff
@@ -1007,6 +1010,211 @@ def prune(cost):
...
+           In KDD'02: Proceedings of the Eighth ACM SIGKDD
+           International Conference on Knowledge Discovery and Data Mining,
+           pp. 538--543. ACM Press, 2002.
+    """
+    # This is the definition from the paper:
+    #
+    # def simrank(G, u, v):
+    #     in_neighbors_u = G.predecessors(u)
+    #     in_neighbors_v = G.predecessors(v)
+    #     scale = C / len(in_neighbors_u) * len(in_neighbors_v)
```
[the remark this answers is not in the document]
Comment (mrecachinas) [reply networkx/algorithms/similarity.py]: The decay/importance factor. Addressed in 6d6ff62.

**Comment by dschult** [comment]
OK... I think we are very close. A few small things to clean up.

- In doc_string definition, should ```C / len(nbrs_u) * len(nbrs_v)``` have parens around the multiplication? (both are in denominator right?)
- At start of example line 1115, you use "from zero to *n* where..." It should be "from 0 to *n-1* where"
- can you split the example on line 1124 to bring it into the 80 char line limit for PEP8.  Maybe like:

```
        >>> lol = [[sim[u][v] for v in sorted(sim[u])] for u in sorted(sim)]
        >>> sim_array = array(lol)
```
lol stands for list-of-list, but any short name could work.

Thanks!

**Comment by mrecachinas** [comment]
@dschult Let me know if there's anything else you want to see changed.

**Comment by dschult** [comment]
Thanks!

**Pull request merged**